### PR TITLE
[OptApp] Relaxed Gradient Projection Algorithm

### PR DESCRIPTION
**Pull request by Igarizza**
**📝 Description**
PR implements relaxed gradient projection method and few tools for it.

### CODE REVIEW

File: applications/OptimizationApplication/python_scripts/algorithms/standardized_rgp_constraint.py [hunk 1]
```diff
@@ -0,0 +1,267 @@
...
+            self.__reference_value = None
+
+        self.__constraint_type = parameters["type"].GetString()
+        if self.__constraint_type in ["<=", "="]:
+            self.__scaling = 1.0
+        elif self.__constraint_type in [">="]:
+            self.__scaling = -1.0
+            if not self.__reference_value is None: self.__reference_value *= -1
+        else:
+            raise RuntimeError(f"Provided \"type\" = {self.__constraint_type} is not supported in constraint response functions. Followings are supported options: \n\t=\n\t<=\n\t<\n\t>=\n\t>")
```
Comment (philbucher) [inline applications/OptimizationApplication/python_scripts/algorithms/standardized_rgp_constraint.py]: consider using an `enum.Enum`:
```py
class ConstraintType(Enum):
    EQUAL = "="
    # ...

if self.constaint_type == Constraint.EQUAL:
    # ....
```

You get the idea, this way it is type safe an easier to read

File: applications/OptimizationApplication/python_scripts/utilities/opt_convergence.py [hunk 2]
```diff
@@ -81,3 +85,97 @@ def GetInfo(self) -> dict:
...
+
+        return self.conv
+
+    def GetInfo(self) -> dict:
+        info = {'type': 'aver_abs_delta',
+                'aver_abs_delta [%]': self.value * 100,
+                'tolerance [%]': self.__tolerance * 100,
+                'tracked_iter': self.__tracked_iter,
+                'status': str("converged" if self.conv else "not converged")}
+        return info
```
Comment (philbucher) [inline applications/OptimizationApplication/python_scripts/utilities/opt_convergence.py:143]: this seems like an ideal application for a `dataclass`:

```py
@dataclass
class Info:
    type_info: str
    aver_abs_delta: float
    # ...
```
this makes it type safe

File: applications/OptimizationApplication/python_scripts/algorithms/standardized_rgp_constraint.py [hunk 3]
```diff
@@ -0,0 +1,267 @@
...
+
+        print(f"RGP Constraint {self.GetResponseName()}:: UpdateBufferSize")
+        print(f"RGP Constraint {self.GetResponseName()}:: CBV = {self.CBV}")
+        print(f"RGP Constraint {self.GetResponseName()}:: BSF = {self.BSF}")
+        print(f"RGP Constraint {self.GetResponseName()}:: BS = {self.BS}")
+        print(f"RGP Constraint {self.GetResponseName()}:: CF = {self.CF}")
+
+    def IsActiveConstrant(self):
+        if self.ComputeW() > 0.0: return True
+        else: return False
```
Comment (philbucher) [inline applications/OptimizationApplication/python_scripts/algorithms/standardized_rgp_constraint.py]:
```suggestion
    def IsActiveConstrant(self):
        return self.ComputeW() > 0.0
```

:sweat_smile:

Comment (Igarizza) [reply applications/OptimizationApplication/python_scripts/algorithms/standardized_rgp_constraint.py]: done

File: applications/OptimizationApplication/python_scripts/algorithms/standardized_rgp_constraint.py (hunk 1 above)
Comment (Igarizza) [reply applications/OptimizationApplication/python_scripts/algorithms/standardized_rgp_constraint.py]: works nice, thnx

File: applications/OptimizationApplication/python_scripts/algorithms/algorithm_relaxed_gradient_projection.py [hunk 4]
```diff
@@ -0,0 +1,317 @@
...
+
+    @time_decorator()
+    def Finalize(self):
+        self.master_control.Finalize()
+        self.__objective.Finalize()
+        CallOnAll(self.__constraints_list, StandardizedRGPConstraint.Finalize)
+
+    @time_decorator()
+    def ComputeSearchDirection(self, obj_grad: KratosOA.CollectiveExpression, constr_grad: 'list[KratosOA.CollectiveExpression]', w_r: Kratos.Vector, w_c: Kratos.Vector) -> KratosOA.CollectiveExpression:
+        active_constraints_list = [self.__constraints_list[i] for i in range(len(self.__constraints_list)) if self.__constraints_list[i].IsActiveConstrant()]
```
Comment (sunethwarna) [inline applications/OptimizationApplication/python_scripts/algorithms/algorithm_relaxed_gradient_projection.py]:
```suggestion
        active_constraints_list = [constraint for constraint in self.__constraints_list if constraint.IsActiveConstrant()]
```

File: applications/OptimizationApplication/python_scripts/algorithms/algorithm_relaxed_gradient_projection.py [hunk 5]
```diff
@@ -0,0 +1,317 @@
...
+            result[i] = KratosOA.ExpressionUtils.InnerProduct(collective_list_item, other_collective)
+        return result
+
+    def __CollectiveListVectorProduct(self, collective_list: 'list[KratosOA.CollectiveExpression]', vector: Kratos.Vector) -> KratosOA.CollectiveExpression:
+        if len(collective_list) != vector.Size():
+            raise RuntimeError(f"Collective list size and vector size mismatch. [ Collective list size = {len(collective_list)}, vector size = {vector.Size()}]")
+        if len(collective_list) == 0:
+            raise RuntimeError("Collective lists cannot be empty.")
+
+        result = collective_list[0] * 0.0
```
Comment (sunethwarna) [inline applications/OptimizationApplication/python_scripts/algorithms/algorithm_relaxed_gradient_projection.py]: I would move these to ExpressionUtils. This is being copy pasted several times.

File: applications/OptimizationApplication/python_scripts/algorithms/algorithm_relaxed_gradient_projection.py [hunk 6]
```diff
@@ -0,0 +1,317 @@
...
+
+    @time_decorator()
+    def Solve(self):
+        while not self.converged:
+            with OptimizationAlgorithmTimeLogger("Gradient Projection",self._optimization_problem.GetStep()):
+                self._InitializeIteration()
+
+                self.__obj_val = self.__objective.CalculateStandardizedValue(self.__control_field)
+                obj_info = self.__objective.GetInfo()
+                self.algorithm_data.GetBufferedData()["std_obj_value"] = obj_info["value"]
```
Comment (sunethwarna) [inline applications/OptimizationApplication/python_scripts/algorithms/algorithm_relaxed_gradient_projection.py]: As far as I see, the `obj_info["value"]` only returns the objective value, not the standardized objective value. Could you check it?

File: applications/OptimizationApplication/python_scripts/algorithms/algorithm_relaxed_gradient_projection.py [hunk 7]
```diff
@@ -0,0 +1,317 @@
...
+                self.Output()
+
+                self.UpdateControl()
+
+                self.converged = self.__convergence_criteria.IsConverged()
+
+                if self.converged:
+                    for constraint in self.__constraints_list:
+                        if not constraint.IsSatisfied():
+                            self.converged = False
```
Comment (sunethwarna) [inline applications/OptimizationApplication/python_scripts/algorithms/algorithm_relaxed_gradient_projection.py]:
```suggestion
                self.converged &= all([c.IsSatisfied() for c in self.__constraints_list])
```

File: applications/OptimizationApplication/python_scripts/algorithms/standardized_rgp_constraint.py [hunk 8]
```diff
@@ -0,0 +1,276 @@
...
+from KratosMultiphysics.OptimizationApplication.utilities.logger_utilities import TimeLogger
+from enum import Enum
+
+
+class ConstraintType(Enum):
+    EQUAL = "="
+    LOWER = "<"
+    LOWEREQ = "<="
+    GREATER = ">"
+    GREATEREQ = ">="
```
Comment (sunethwarna) [inline applications/OptimizationApplication/python_scripts/algorithms/standardized_rgp_constraint.py:17]: Nice.... We should do the same for the other algorithms as well.

File: applications/OptimizationApplication/python_scripts/utilities/opt_convergence.py (hunk 2 above)
Comment (sunethwarna) [reply applications/OptimizationApplication/python_scripts/utilities/opt_convergence.py:143]: We need to have properly defined interface class for `ConvergenceCriteria`, in this case the data class would not be possible ryt?

**Review by sunethwarna — COMMENTED** [review]
Nice work @Igarizza. I have following suggestion apart from the in plce suggestions.
1. Can we move the "shell_test.mdpa" to the "mdpas" folder. Because, you use the same mdpa in some other algorithm tests as well.  (Same for "primal_parameters.json, StructuralMateirals.json")
2. Would it be possible to derrive `standardized_rgp_constraint` from `standardized_contraint`?.  Because, there are some duplicated codes there. Or, we can move all of them to utils and use it in both places.

File: applications/OptimizationApplication/python_scripts/utilities/opt_convergence.py (hunk 2 above)
Comment (philbucher) [reply applications/OptimizationApplication/python_scripts/utilities/opt_convergence.py:143]: a dataclass is just a class in the end ...
you can add methods and everything to it

Having said this, if you want to leave the `status` var, it should probably also be an enum => `CONVERGED` & `NOT_CONVERGED`

File: applications/OptimizationApplication/python_scripts/algorithms/algorithm_relaxed_gradient_projection.py (hunk 6 above)
Comment (Igarizza) [reply applications/OptimizationApplication/python_scripts/algorithms/algorithm_relaxed_gradient_projection.py]: you are right, fixed.

**Comment by Igarizza** [comment]
@sunethwarna I can inherit rgp constraint from standart constraint, but they have different settings, hence, I need to create dummy standart constraint settings to retrofit rgp settings. It doesn't look good and can be very confusing.

**Review by sunethwarna — DISMISSED** [review]

**Review by sunethwarna — APPROVED** [review]

**Pull request merged**